kobo/apps/accounts/adapter.py

```diff
@@ -1,10 +1,52 @@
 from allauth.account.adapter import DefaultAccountAdapter
 from allauth.account.forms import SignupForm
 from constance import config
+from django.conf import settings
+from django.contrib.auth import REDIRECT_FIELD_NAME
+from django.contrib.sites.shortcuts import get_current_site
 from django.db import transaction
+from django.shortcuts import resolve_url
+from django.template.response import TemplateResponse
+from trench.utils import get_mfa_model, user_token_generator
+
+from .mfa.forms import MfaTokenForm
+from .mfa.views import MfaTokenView
 
 
 class AccountAdapter(DefaultAccountAdapter):
+
+    def pre_login(self, request, user, **kwargs):
+
+        auth_method = (
+            get_mfa_model()
+            .objects.filter(is_active=True, user=user)
+            .first()
+        )
+        if auth_method:
+            ephemeral_token_cache = user_token_generator.make_token(user)
+            mfa_token_form = MfaTokenForm(
+                initial={'ephemeral_token': ephemeral_token_cache}
+            )
+            next_url = kwargs.get(
+                'redirect_url', resolve_url(settings.LOGIN_REDIRECT_URL)
+            )
+            current_site = get_current_site(request)
+            context = {
+                REDIRECT_FIELD_NAME: next_url,
+                'site': current_site,
+                'site_name': current_site.name,
```

Comment:
Can we avoid introducing the Django sites framework? It's in kobocat and it's annoying; it seems like we've gotten this far in kpi without using it. I see that `django.contrib.sites` has _not_ been added to `INSTALLED_APPS`, meaning we can't actually use the `Site` model; is this invocation of the sites framework a fancy way to call `request.get_host()` for `site_name`[^1], or does the template do more stuff with the `site` object?

[^1]:
    ```
    In [1]: from django.contrib.sites.shortcuts import get_current_site

    In [2]: class C: pass

    In [3]: req = C()

    In [4]: get_current_site(req).name
    ---------------------------------------------------------------------------
    <snip>
    AttributeError: 'C' object has no attribute 'get_host'

    In [5]: req.get_host = lambda: 'foo.com'

    In [6]: get_current_site(req).name
    Out[6]: 'foo.com'
    ```

Reply:
Because it's inheriting for `LoginView`, I thought it was required because of the [context passed to the template ](https://github.com/django/django/blob/main/django/contrib/auth/views.py#L110-L121).

But it seems to work without it.

Reply:
OK. I guess I don't totally understand the relationship with `LoginView`, but since neither `mfa_token.html` nor `account/base.html` use `site` or `site_name`, we should be fine